i3_workspace_names_daemon.py

```diff
@@ -39,21 +38,55 @@ def build_rename(i3, app_icons, delim):
         The rename callback.
     """
     def get_icon_or_name(leaf):
-        window_class = leaf.window_class.lower()
+        window_class = ""
+        if leaf.window_class:
+            window_class = leaf.window_class.lower()
+        else:
+            window_class = leaf.name.lower()
+
         if window_class in app_icons and app_icons[window_class] in icons:
             return icons[app_icons[window_class]]
         else:
             return window_class.lower()
 
     def rename(i3, e):
-        for workspace in i3.get_tree().workspaces():
+        workspaces = i3.get_tree().workspaces()
+        #need to use get_workspaces since the i3 con object doesn't have the visible property for some reason
+        workdicts = i3.get_workspaces()
+        visible =  [workdict['name'] for workdict in workdicts if workdict['visible']]
+        visworkspaces = []
+        focus = [workdict['name'] for workdict in workdicts if workdict['focused']]
+        focusname = None
+        if len(focus) > 0:
+            focus = focus[0]
+        else:
+            focus = None
+
+        commands = []
+        for workspace in workspaces:
             names = [get_icon_or_name(leaf)
                      for leaf in workspace.leaves()]
             names = delim.join(names)
+            newname = ""
             if int(workspace.num) > 0:
-                i3.command('rename workspace "{}" to "{}: {}"'.format(workspace.name, workspace.num, names))
+                newname = "{}: {}".format(workspace.num, names)
             else:
-                i3.command('rename workspace "{}" to "{}"'.format(workspace.name, names))
+                newname = names 
+
+            if workspace.name in visible:
+                visworkspaces.append(newname)
+            if workspace.name == focus:
+                focusname = newname
+
+            commands.append('rename workspace "{}" to "{}"'.format(workspace.name, newname))
+
+        for workspace in visworkspaces + [focusname]:
+            commands.append('workspace {}'.format(workspace))
+
+        #we have to join all the activate workspaces commands into one or the order
```

Comment:
ah, makes sense.

Reply:
Yeah this is the solution to #6. I assume there is some underlying async in i3ipc that can allow the order of command to be changed. The simple solution is to just issue one long command at the end of rename. I should probably check and make sure there is no command length limit in i3ipc or i3 itself.